File: backend/workout-service/src/workout/workout.service.ts
import { Injectable, NotFoundException } from '@nestjs/common';
import { InjectRepository } from '@nestjs/typeorm';
import { Repository } from 'typeorm';
import { WorkoutSession } from './workout-session.entity';
import { WorkoutExercise } from './workout-exercise.entity';
import { estimateSessionCalories, estimateExerciseCalories } from './calorie.service';

@Injectable()
export class WorkoutService {
  constructor(
    @InjectRepository(WorkoutSession) private sessions: Repository<WorkoutSession>,
    @InjectRepository(WorkoutExercise) private exercises: Repository<WorkoutExercise>,
  ) {}

  async createSession(userId: string, data: any, userWeightKg?: number): Promise<WorkoutSession> {
    const exercisesData = data.exercises || [];
    const session = this.sessions.create({
      user_id: userId,
      title: data.title,
      workout_type: data.workout_type,
      started_at: data.started_at ? new Date(data.started_at) : new Date(),
      duration_minutes: data.duration_minutes,
      is_public: data.is_public ?? true,
      notes: data.notes,
    });
    if (session.duration_minutes) {
      session.total_calories = estimateSessionCalories(session.workout_type, session.duration_minutes, userWeightKg);
    }
    await this.sessions.save(session);

    let totalCal = 0;
    for (const ex of exercisesData) {
      const { calories, met } = estimateExerciseCalories(
        ex.exercise_name, ex.sets, ex.reps, ex.weight_kg, ex.duration_sec, userWeightKg,
      );
      const exercise = this.exercises.create({ ...ex, session_id: session.id, calories, met_value: met });
      await this.exercises.save(exercise);
      totalCal += calories;
    }
    if (totalCal > 0 && !session.duration_minutes) {
      session.total_calories = Math.round(totalCal * 1.1 * 100) / 100;
      await this.sessions.save(session);
    }
    return this.sessions.findOne({ where: { id: session.id }, relations: ['exercises'] });
  }

  async addExercise(sessionId: string, userId: string, data: any, userWeightKg?: number) {
    const session = await this.sessions.findOne({ where: { id: sessionId } });
    if (!session || session.user_id !== userId) throw new NotFoundException('Session not found');
    const { calories, met } = estimateExerciseCalories(
      data.exercise_name, data.sets, data.reps, data.weight_kg, data.duration_sec, userWeightKg,
    );
    const exercise = this.exercises.create({ ...data, session_id: sessionId, calories, met_value: met });
    await this.exercises.save(exercise);
    const cur = parseFloat(session.total_calories as any) || 0;
    session.total_calories = Math.round((cur + calories) * 100) / 100;
    await this.sessions.save(session);
    return exercise;
  }

  getSession(id: string) {
    return this.sessions.findOne({ where: { id }, relations: ['exercises'] });
  }

  getUserSessions(userId: string, limit = 50, offset = 0) {
    return this.sessions.find({ where: { user_id: userId }, order: { started_at: 'DESC' }, take: limit, skip: offset });
  }

  getPublicSessionsByUser(userId: string) {
    return this.sessions.find({ where: { user_id: userId, is_public: true }, order: { started_at: 'DESC' } });
  }

  async deleteSession(id: string, userId: string): Promise<boolean> {
    const s = await this.sessions.findOne({ where: { id } });
    if (!s || s.user_id !== userId) return false;
    await this.sessions.remove(s);
    return true;
  }

  async getStats(userId: string) {
    const result = await this.sessions.createQueryBuilder('s')
      .select('COUNT(*)', 'total_sessions')
      .addSelect('SUM(s.total_calories)', 'total_calories')
      .addSelect('SUM(s.duration_minutes)', 'total_minutes')
      .where('s.user_id = :userId', { userId })
      .getRawOne();
    return result;
  }
}

File: backend/workout-service/app/service/workout_service.py
from typing import List, Optional
from uuid import UUID
from datetime import datetime
from sqlalchemy.orm import Session

from app.models.workout import WorkoutSession, WorkoutExercise
from app.repository.workout_repository import WorkoutRepository
from app.service.calorie_service import (
    estimate_session_calories,
    estimate_exercise_calories,
    DEFAULT_WEIGHT_KG,
)


class WorkoutService:
    def __init__(self, db: Session):
        self.repo = WorkoutRepository(db)

    def create_session(self, user_id: UUID, data: dict, user_weight_kg: float = None) -> WorkoutSession:
        exercises_data = data.pop("exercises", []) or []

        session = WorkoutSession(
            user_id=user_id,
            title=data["title"],
            workout_type=data["workout_type"],
            started_at=data.get("started_at") or datetime.utcnow(),
            duration_minutes=data.get("duration_minutes"),
            is_public=data.get("is_public", True),
            notes=data.get("notes"),
        )

        if session.duration_minutes:
            session.total_calories = estimate_session_calories(
                session.workout_type, session.duration_minutes, user_weight_kg
            )

        session = self.repo.create_session(session)

        total_calories = 0.0
        for ex_data in exercises_data:
            exercise, cal = self._build_exercise(session.id, ex_data, user_weight_kg)
            self.repo.add_exercise(exercise)
            total_calories += cal

        if total_calories > 0 and not session.duration_minutes:
            session.total_calories = round(total_calories * 1.1, 2)  # +10% overhead
            self.repo.update_session(session)

        return self.repo.find_session_by_id(session.id)

    def add_exercise(self, session_id: UUID, user_id: UUID, data: dict, user_weight_kg: float = None) -> Optional[WorkoutExercise]:
        session = self.repo.find_session_by_id(session_id)
        if not session or session.user_id != user_id:
            return None

        exercise, cal = self._build_exercise(session_id, data, user_weight_kg)
        exercise = self.repo.add_exercise(exercise)

        # Update session total_calories
        current = float(session.total_calories or 0)
        session.total_calories = round(current + cal, 2)
        self.repo.update_session(session)

        return exercise

    def _build_exercise(self, session_id: UUID, data: dict, user_weight_kg: float = None):
        calories, met = estimate_exercise_calories(
            exercise_name=data.get("exercise_name", "general"),
            sets=data.get("sets"),
            reps=data.get("reps"),
            weight_kg=data.get("weight_kg"),
            duration_sec=data.get("duration_sec"),
            user_weight_kg=user_weight_kg,
        )
        exercise = WorkoutExercise(
            session_id=session_id,
            exercise_name=data["exercise_name"],
            sets=data.get("sets"),
            reps=data.get("reps"),
            weight_kg=data.get("weight_kg"),
            duration_sec=data.get("duration_sec"),
            distance_km=data.get("distance_km"),
            calories=calories,
            met_value=met,
        )
        return exercise, calories

    def get_session(self, session_id: UUID) -> Optional[WorkoutSession]:
        return self.repo.find_session_by_id(session_id)

    def get_user_sessions(self, user_id: UUID, limit: int = 50, offset: int = 0) -> List[WorkoutSession]:
        return self.repo.find_sessions_by_user(user_id, limit, offset)

    def get_partner_sessions(self, partner_id: UUID) -> List[WorkoutSession]:
        return self.repo.find_public_sessions_by_user(partner_id)

    def delete_session(self, session_id: UUID, user_id: UUID) -> bool:
        session = self.repo.find_session_by_id(session_id)
        if not session or session.user_id != user_id:
            return False
        self.repo.delete_session(session)
        return True

    def get_stats(self, user_id: UUID) -> dict:
        return self.repo.get_user_stats(user_id)
